### PR TITLE
Setting up redis without password

### LINKED ISSUE #249: Running redis without a password - discussion or PR?
Ahoy! 

I want to run redis in my trusted network, not exposed to the internet, and I think that [the redis docs](http://redis.io/topics/security) imply that that is the primary way that redis was designed to be used. Unfortunately, I think the stable helm chart here does not expose a way to configure redis without a password.

I'm not sure what the process is like in this repo for suggesting changes. Would you all be interested in further discussion about this? I'd be happy to work on a PR for making passwords optional, but I'm not sure that a) that change would be welcome or b) what style of configuration is conventional for turning _off_ security in an official Helm chart.

I hope this makes sense! Please let me know if I can clarify or if I'm misunderstanding something.

### PR DESCRIPTION
**Pull request by gedimin45**
closes #249

### CODE REVIEW

**Comment by lachie83** [comment]
Thanks for the contrib @ged15. Just a quick question on this one. Should we make having a password default with the ability to disable? cc @prydonius @viglesiasce

**Comment by gedimin45** [comment]
I would say the default should be a password-less setup. Based on my experience, Redis is not exposed outside the private network thus authentication is unnecessary and the need for a password is a special case.

**Comment by prydonius** [comment]
It is quite common to run Redis without authentication when not exposed externally, and reading through https://redis.io/topics/security it seems like they strongly suggest only exposing Redis within trusted environments (rather than making it externally available and enabling authentication).

Given the chart does not expose Redis externally by default, a password-less setup should be okay and is probably more in line with what people would expect out-of-the-box.

**Comment by miracle2k** [comment]
Ran into this; I actually very much like that it has the ability to generate a password. I'd like to see that ability persist. It would suggest that there should be the ability to disable this though, possible by setting the password to an empty string.

**Comment by gedimin45** [comment]
If you ask me, generating passwords can be done using a script by the operations person that is installing this and there is no need to keep this feature in the charts. The way I see it, generating random passwords has minimal utility (Redis is used without a password usually) and can be easily substituted by an outside tool.
That being said, what do the maintainers think? If we decide to keep this feature, I can modify the chart to generate the password by default adding a feature to disable authentication.
@lachie83?

**Comment by viglesiasce** [comment]
I would like us to keep the current default and allow users to set a value that disables authentication.

**Comment by miracle2k** [comment]
It's actually not easy to manually set a redis password, because it seems the image only accepts the password on the first start, according to the docs.

It's true that redis is frequently used without a password, including by me, but I don't see how this is more secure than running an internal MySQL without a password. I like that the chart encourages me to use one by making that easy/the default.

**Comment by prydonius** [comment]
@ged15 would you be willing to update this PR to add an option for disabling the password, but keeping the current default?

**Comment by gedimin45** [comment]
@prydonius yes. Will try to take a look at this during the weekend.

**Comment by lachie83** [comment]
@ged15 This also needs a rebase.

**Comment by prydonius** [comment]
Moving to Milestone 2

**Comment by lachie83** [comment]
@k8s-bot ok to test

**Comment by lachie83** [comment]
Rebased

File: stable/redis/values.yaml [hunk 1]
```diff
@@ -9,7 +9,6 @@ image: bitnami/redis:3.2.6-r2
 imagePullPolicy: IfNotPresent
 
 ## Redis password
-## Defaults to a random 10-character alphanumeric string if not set
 ## ref: https://github.com/bitnami/bitnami-docker-redis#setting-the-server-password-on-first-run
 ##
 # redisPassword:
```
Comment (lachie83) [inline stable/redis/values.yaml:14]: Please keep the default to generate a password

**Review by lachie83 — CHANGES_REQUESTED** [review]

**Comment by lachie83** [comment]
@ged15 Can you please provide a status update?

**Comment by gajus** [comment]
> @ged15 Can you please provide a status update?

Looking at @ged15 GitHub publicly activity, its unlikely there will be a follow up.

Can you please update the PR and merge it? I am assuming the default setting is the only outstanding issue.

**Comment by sstarcher** [comment]
@ged15 could you finish the changes on this?  If not I would be happy to do it for you.

**Comment by gedimin45** [comment]
Sorry for the inactivity, sorting out some personal stuff.
@sstarcher I would really like to finish this but realistically that won't happen any time soon. If you want to pick this up, go ahead!

**Comment by sstarcher** [comment]
@lachie83 I looked into this issue more.  Currently with the use of rand in the chart itself will cause a helm upgrade to regenerate the password causing any pods that were referencing it to no longer connect to redis.

This certainly effects the `sensu` chart that exists.  The current only way around this is for the user to specify the password.  I considered using .Release.IsUpgrade around the redis secrets, but this of course causes the secret to be fully deleted which is also not helpful.  

I'm not currently seeing any good options outside of defaulting redis to password less.  I consider the current option of using rand to be a bad option due to how upgrades work.

**Comment by viglesiasce** [comment]
@sstarcher should the NOTES.txt tell people to use --set redisPassword=\<auto generated password\> ?

**Comment by sstarcher** [comment]
@viglesiasce I believe the current way these charts functions is flawed.  As users can't upgrade the chart without the password changing.  The best course of action until this is resolved would be to default the redis configuration to no password, but we have a larger Helm question of how to handle generated information without regenerating that information on upgrade.

**Comment by viglesiasce** [comment]
Sorry I mean that:

1) the chart should tell them that they must record the password for upgrades
2) We block the regeneration of the password based on the release number
3) tell them to enter their previous password or else things will break on upgrade

**Comment by sstarcher** [comment]
@viglesiasce that would certainly work, but the UX would be sub-par.  To upgrade the user should not have to know the current password.

**Comment by viglesiasce** [comment]
@sstarcher what do you suggest?

**Comment by sstarcher** [comment]
I don't currently have a great suggestion, but the current mechanism that is being recommended to people is a bad practice that they won't notice until they attempt to upgrade.  Helm does not have any in-built solution that would make this easy.

**Comment by ianburrell** [comment]
The [docs](https://github.com/kubernetes/helm/blob/master/docs/charts_tips_and_tricks.md#automatically-roll-deployments-when-configmaps-or-secrets-change) mention a technique using annotations to update deployments when the configmap or secrets change. That could be used to fix problems where changing the password isn't propagated.

There is still the problem of generated password changing with every update and causing unnecessary updates. It seems like this issue needs to be addressed by Helm because it is a problem for lots of software.

**Comment by ianburrell** [comment]
I have a branch that keeps random default password but supports not using a password. It does this by adding a value, usePassword, that defaults to "true". But setting it to "false" disables the secret and REDIS_PASSWORD.

**Comment by sstarcher** [comment]
@ianburrell nice I had not seen that new addition to helm.  @ianburrell mind putting in a PR with that change.  That sounds like what this PR is trying to accomplish.

**Comment by viglesiasce** [comment]
ping @ianburrell

**Comment by ianburrell** [comment]
@viglesiasce I put in #1061. Working on getting CLA signed.

**Comment by prydonius** [comment]
closing in favour of #1061

**Pull request closed without merging**